quickwit-search/src/root.rs

```diff
@@ -88,48 +212,39 @@ pub async fn root_search(
     search_request_with_offset_0.max_hits += search_request.start_offset;
 
     // Perform the query phase.
-    let mut leaf_search_handles: Vec<JoinHandle<Result<LeafSearchResult, tonic::Status>>> =
-        Vec::new();
-    for (search_client, jobs) in assigned_leaf_search_jobs.iter() {
-        let leaf_search_request = LeafSearchRequest {
-            search_request: Some(search_request_with_offset_0.clone()),
-            split_ids: jobs.iter().map(|job| job.split.clone()).collect(),
-        };
+    let result_per_node_addr = execute_search(
+        &assigned_leaf_search_jobs,
+        search_request_with_offset_0.clone(),
+    )
+    .await?;
+
+    analyze_errors(&result_per_node_addr);
 
-        // TODO wrap search clients with some info like their socketaddr to be able to log useful debug information
-        debug!(leaf_search_request=?leaf_search_request, "Leaf node search.");
-        let mut search_client_clone = search_client.clone();
-        let handle = tokio::spawn(async move {
-            search_client_clone
-                .client()
-                .leaf_search(leaf_search_request)
-                .await
-                .map(|resp| resp.into_inner())
-        });
-        leaf_search_handles.push(handle);
-    }
-    let leaf_search_responses = futures::future::try_join_all(leaf_search_handles).await?; //< An error here means that the tokio task panicked... Not that the grpc erorred.
-                                                                                           //< The latter is handled later.
     let index_metadata = metastore.index_metadata(&search_request.index_id).await?;
     let index_config = index_metadata.index_config;
     let collector = make_collector(index_config.as_ref(), search_request);
 
     // Find the sum of the number of hits and merge multiple partial hits into a single partial hits.
     let mut leaf_search_results = Vec::new();
-    for leaf_search_response in leaf_search_responses {
+    for (_addr, leaf_search_response) in result_per_node_addr.iter() {
         match leaf_search_response {
             Ok(leaf_search_result) => {
                 debug!(leaf_search_result=?leaf_search_result, "Leaf search result.");
                 leaf_search_results.push(leaf_search_result)
             }
             Err(grpc_error) => {
-                let leaf_search_error = parse_grpc_error(&grpc_error);
+                let leaf_search_error = parse_grpc_error(&grpc_error.status);
                 error!(error=?grpc_error, "Leaf request failed");
                 // TODO list failed leaf nodes and retry.
                 return Err(leaf_search_error);
             }
         }
     }
+
+    let leaf_search_results = result_per_node_addr
+        .into_iter()
+        .map(|(_addr, results)| results.unwrap())
+        .collect_vec();
```

Comment:
What is the point of this? It seems like you computed that above already...

(told you shadowing is evil :) )

Reply:
oops this is some residue

(... no harm done by shadowing ... :) )